src/coreclr/src/System.Private.CoreLib/src/System/ActivatorCache.CoreCLR.cs

```diff
@@ -0,0 +1,164 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System
+{
+    internal sealed partial class RuntimeType
+    {
+        /// <summary>
+        /// A cache which allows optimizing <see cref="Activator.CreateInstance"/>,
+        /// <see cref="RuntimeType.CreateInstanceDefaultCtor"/>, and related APIs.
+        /// </summary>
+        private sealed unsafe class ActivatorCache
+        {
+            // The managed calli to the newobj allocator, plus its first argument (MethodTable*).
+            // In the case of the COM allocator, first arg is ComClassFactory*, not MethodTable*.
+            private readonly delegate*<IntPtr, object?> _pfnAllocator;
+            private readonly IntPtr _allocatorFirstArg;
+
+            // The managed calli to the parameterless ctor, taking "this" (as object) as its first argument.
+            // For value type ctors, we'll point to a special unboxing stub.
+            private readonly delegate*<object?, void> _pfnCtor;
+
+#if DEBUG
+            private readonly WeakReference<RuntimeType> _originalRuntimeType; // don't prevent the RT from being collected
```

Comment:
What is the weak reference trying to achieve? The GC should be able to handle the cycles just fine.

Reply:
I think it's left over from when I was experimenting with COM support, trying not to root the `RuntimeType` instances that had CLSIDs hanging off their sync blocks. But yeah, this isn't needed any more.